creusot/src/validate_terminates.rs

```diff
@@ -196,28 +196,34 @@ pub(crate) fn validate_terminates(ctx: &mut TranslationCtx) {
                         // Let's try to find if this specific invocation can be specialized to a known implementation
                         let actual_clause = EarlyBinder::bind(clause.skip_binder())
                             .instantiate(ctx.tcx, generic_args);
+                        let param_env = ctx.tcx.param_env(caller_def_id);
                         let obligation = TraitObligation::new(
                             ctx.tcx,
                             ObligationCause::dummy(),
-                            ctx.tcx.param_env(caller_def_id),
+                            param_env,
                             actual_clause,
                         );
                         let infer_ctx = ctx.infer_ctxt().intercrate(true).build();
                         let mut selection_ctx = SelectionContext::new(&infer_ctx);
-                        let impl_def_id = match selection_ctx.select(&obligation) {
-                            Ok(Some(source)) => match source {
-                                rustc_infer::traits::ImplSource::UserDefined(source) => {
-                                    Some(source.impl_def_id)
-                                }
-                                rustc_infer::traits::ImplSource::Param(_) => {
-                                    // FIXME: we take the conservative approach here, but what does this case actually mean ?
-                                    None
-                                }
-                                // Used for marker traits (no functions anyway) and trait object/unsized variables (we really don't know what they can call)
-                                rustc_infer::traits::ImplSource::Builtin(_, _) => None,
-                            },
-                            Ok(None) => None,
-                            Err(_) => None,
+                        let impl_def_id = if actual_clause.trait_ref.has_param() {
+                            // else this crashes the trait selection.
+                            None
+                        } else {
+                            match selection_ctx.select(&obligation) {
+                                Ok(Some(source)) => match source {
+                                    rustc_infer::traits::ImplSource::UserDefined(source) => {
+                                        Some(source.impl_def_id)
+                                    }
+                                    rustc_infer::traits::ImplSource::Param(_) => {
+                                        // FIXME: we take the conservative approach here, but what does this case actually mean ?
```

Comment:
This case means that the impl comes from a bound on a parameter, I think we have to be very conservative here since a clever instantiation of that generic & bound is exactly how you might cheat the detection.

Reply:
Actually, I don't think there is any risk here, as long as we add a dependency at the instantiation site. Think of polymorphism in system F (which is terminating, as you know) : in system F, we add no dependency when using a type variable.

Reply:
I see what you mean, but I think at the very least his module would then benefit from more extensive description of the criteria, what is up there is quite good but this is fairly sophisticated reasoning.

Reply:
I don't understand: do you simply complain about the missing doc?

Reply:
Ok, thanks for the explanation!
Anyway this code wasn't working with the latest rustc, and I've taken the approach of using `TyCtxt::resolve_instance` here, which seems more robust and simpler to read.